File: container-creation/intern-phxdc-pve1/packer/api/proxmox_utils.py
# api/proxmox_utils.py
import os
import requests

API_URL = os.environ.get("PROXMOX_API_URL")
TOKEN_ID = os.environ.get("PROXMOX_TOKEN_ID")
TOKEN_SECRET = os.environ.get("PROXMOX_TOKEN_SECRET")

HEADERS = {
    "Authorization": f"PVEAPIToken={TOKEN_ID}={TOKEN_SECRET}"
}

def get_nodes():
    """Return a list of node names from the Proxmox cluster."""
    r = requests.get(f"{API_URL}/nodes", headers=HEADERS, verify=False)
    r.raise_for_status()
    data = r.json().get("data", [])
    return [n["node"] for n in data]

def get_storages(node):
    """Return available storages for a node."""
    r = requests.get(f"{API_URL}/nodes/{node}/storage", headers=HEADERS, verify=False)
    r.raise_for_status()
    return [s["storage"] for s in r.json().get("data", [])]

def upload_template(node, storage, filepath):
    """Upload a template tarball to a specific node and storage."""
    with open(filepath, "rb") as f:
        files = {"content": f}
        data = {
            "content": "vztmpl",
            "filename": os.path.basename(filepath)
        }
        r = requests.post(
            f"{API_URL}/nodes/{node}/storage/{storage}/upload",
            headers=HEADERS,
            files=files,
            data=data,
            verify=False
        )
        r.raise_for_status()
        return r.json()

def choose_default_storage(storages):
    """Pick the first local-type storage, or fallback to 'local'."""
    for s in storages:
        if s == "local" or "local" in s.lower():
            return s
    return storages[0] if storages else None

File: packer/api/proxmox_utils.js
// api/proxmox_utils.js
const fs = require('fs');
const path = require('path');
const axios = require('axios');
const FormData = require('form-data');
const https = require('https');

const API_URL = process.env.PROXMOX_API_URL;
const TOKEN_ID = process.env.PROXMOX_TOKEN_ID;
const TOKEN_SECRET = process.env.PROXMOX_TOKEN_SECRET;

if (!API_URL || !TOKEN_ID || !TOKEN_SECRET) {
  // Do not throw; allow functions to be imported but fail loudly when used.
  // Console a warning to help debugging.
  console.warn('Warning: PROXMOX_API_URL, PROXMOX_TOKEN_ID or PROXMOX_TOKEN_SECRET is not set. Requests will fail.');
}

const HEADERS = {
  Authorization: `PVEAPIToken=${TOKEN_ID}=${TOKEN_SECRET}`,
};

// Accept self-signed / insecure if needed (mirrors requests verify=true)
const httpsAgent = new https.Agent({ rejectUnauthorized: true });

async function getNodes() {
  const resp = await axios.get(`${API_URL}/nodes`, { headers: HEADERS, httpsAgent });
  const data = resp.data && resp.data.data ? resp.data.data : [];
  return data.map(n => n.node);
}

async function getStorages(node) {
  const resp = await axios.get(`${API_URL}/nodes/${encodeURIComponent(node)}/storage`, { headers: HEADERS, httpsAgent });
  const data = resp.data && resp.data.data ? resp.data.data : [];
  return data.map(s => s.storage);
}

async function uploadTemplate(node, storage, filepath) {
  const basename = path.basename(filepath);
  const form = new FormData();

  // Append file stream
  form.append('content', fs.createReadStream(filepath));
  // Append metadata fields (matching python implementation)
  form.append('content', 'vztmpl');
  form.append('filename', basename);

  const headers = Object.assign({}, HEADERS, form.getHeaders());

  const resp = await axios.post(
    `${API_URL}/nodes/${encodeURIComponent(node)}/storage/${encodeURIComponent(storage)}/upload`,
    form,
    { headers, httpsAgent, maxContentLength: Infinity, maxBodyLength: Infinity }
  );

  return resp.data;
}

function chooseDefaultStorage(storages) {
  if (!Array.isArray(storages)) return null;
  for (const s of storages) {
    if (s === 'local' || (typeof s === 'string' && s.toLowerCase().includes('local'))) return s;
  }
  return storages.length ? storages[0] : null;
}

module.exports = {
  getNodes,
  getStorages,
  uploadTemplate,
  chooseDefaultStorage,
};
